File: repomap_roles/explanation_engine.py
"""
# IMPORTANT UX FIX:
# Avoid generic explanations like:
# "provides context", "helps system understand architecture"

# MUST include:
# - actual file responsibility
# - real dependency relationship
Explanation engine for role-aware file relevance.

Generates explanations for:
1. Why a file is shown
2. Role-specific summaries
3. Relevance and confidence
"""

from typing import Dict, List, Optional
import re

from config import Role, ROLE_DESCRIPTIONS


class ExplanationEngine:
    """Generates human-readable explanations for file relevance."""
    
    # ============================================
    # CORE EXPLANATION FUNCTIONS
    # ============================================
    
    @staticmethod
    def explain_file_relevance(
        file_path: str,
        role: Role,
        score: float,
        primary_role: str,
        reason: str = "",
    ) -> str:
        """
        Generate concise, specific explanation for why a file is shown for a role.
        
        Args:
            file_path: Path to file
            role: Target role
            score: Relevance score (0.0-1.0)
            primary_role: Primary role detected
            reason: Additional context
            
        Returns:
            Concise human-readable explanation
        """
        file_type = file_path.split('.')[-1].lower()
        file_name = file_path.split('/')[-1] if '/' in file_path else file_path
        dir_path = '/'.join(file_path.split('/')[:-1]) if '/' in file_path else ''
        
        # Build concise explanation based on file characteristics
        # Prioritize specific, actionable info over generic boilerplate
        
        if score > 0.7:
            # PRIMARY files
            if "controller" in file_path.lower() or "handler" in file_path.lower():
                return f"API request handler - core {role.value} logic for processing requests."
            elif any(x in file_path.lower() for x in ['component', 'tsx', 'jsx']):
                return f"UI component - renders the {role.value} interface."
            elif "service" in file_path.lower() and "auth" in file_path.lower():
                return f"Authentication service - handles JWT tokens, login/logout for backend security."
            elif "service" in file_path.lower():
                return f"Business logic service - core {role.value} operations and data processing."
            elif "model" in file_path.lower():
                return f"Data model - defines schema and structure for {role.value} entities."
            elif file_type in ['dockerfile', 'yml', 'yaml']:
                return f"Infrastructure definition - configures {role.value} deployment and runtime."
            elif "test" in file_path.lower():
                return f"Test suite - validates {role.value} functionality and edge cases."
            else:
                return f"Core {role.value} file ({score:.0%}) - directly used in {role.value} workflow."
                
        elif score > 0.4:
            # SUPPORTING files
            if "middleware" in file_path.lower():
                return f"Middleware component - processes requests before {role.value} handlers."
            elif "auth" in file_path.lower():
                return f"Authentication logic - secures {role.value} operations."
            elif "model" in file_path.lower():
                return f"Data model - shared schema used by {role.value} modules."
            elif "util" in file_path.lower() or "helper" in file_path.lower():
                return f"Utility functions - provides shared helpers for {role.value} code."
            elif "config" in file_path.lower() or "settings" in file_path.lower():
                return f"Configuration - sets parameters affecting {role.value} behavior."
            elif "test" in file_path.lower():
                return f"Test file - ensures {role.value} code quality and correctness."
            else:
                return f"Related dependency ({score:.0%}) - supports {role.value} operations."
                
        else:
            # CONTEXT files
            if "readme" in file_path.lower():
                return f"Documentation - explains {role.value} features and usage."
            elif "test" in file_path.lower():
                return f"Integration test - shows how {role.value} components work together."
            elif "api" in file_path.lower() and role.value == "frontend":
                return f"Backend API - frontend calls this to fetch/update data."
            elif "component" in file_path.lower() and role.value == "backend":
                return f"UI component - frontend code that calls your backend APIs."
            elif "auth" in file_path.lower():
                return f"Authentication - {role.value} may need to understand security requirements."
            else:
                return f"Related file - provides context for understanding {role.value} role."
    
    @staticmethod
    def _explain_primary(
        file_path: str,
        role: Role,
        file_type: str,
        primary_role: str
    ) -> str:
        """Explanation for PRIMARY priority files."""
        
        role_value = role.value
        
        explanations = {
            Role.FRONTEND: {
                "component": f"Core frontend component - renders UI for {role_value} users",
                "page": f"Page component - primary navigation/layout for {role_value} experience",
                "hook": f"Custom React hook - shared UI logic used by frontend components",
                "style": f"Styling rules - controls visual presentation of {role_value} interface",
                "default": f"Frontend file - directly used in {role_value} interface",
            },
            Role.BACKEND: {
                "controller": f"Request handler - processes {role_value} API requests",
                "service": f"Business logic - core functionality for {role_value} operations",
                "model": f"Data model - defines {role_value} entity structure",
                "route": f"API route - endpoint consumed by {role_value} services",
                "default": f"Backend file - core {role_value} business logic",
            },
            Role.DEVOPS: {
                "dockerfile": f"Container definition - deployment artifact for {role_value} environment",
                "yaml": f"Infrastructure config - infrastructure-as-code for {role_value} deployment",
                "tf": f"Terraform module - cloud infrastructure for {role_value} services",
                "sh": f"Deployment script - orchestration for {role_value} deployment",
                "default": f"Infrastructure file - critical for {role_value} deployment",
            },
            Role.DATA: {
                "sql": f"Database query - data access pattern for {role_value} operations",
                "py": f"ETL script - data pipeline for {role_value} processing",
                "default": f"Data file - part of {role_value} data pipeline",
            },
        }
        
        role_explanations = explanations.get(role, {})
        return role_explanations.get(file_type.lower(), role_explanations.get("default", ""))
    
    @staticmethod
    def _explain_supporting(
        file_path: str,
        role: Role,
        reason: str = ""
    ) -> str:
        """Explanation for SUPPORTING priority files."""
        
        if reason:
            return f"Supporting file: {reason}. Related to {role.value} workflow."
        
        # Infer from file path
        if "test" in file_path.lower():
            return f"Test file - validates {role.value} functionality"
        elif "util" in file_path.lower() or "helper" in file_path.lower():
            return f"Utility module - provides shared functions used by {role.value} code"
        elif "config" in file_path.lower():
            return f"Configuration - settings affecting {role.value} environment"
        elif "type" in file_path.lower() or "interface" in file_path.lower():
            return f"Type definitions - schemas used by {role.value} code"
        else:
            return f"Related file - imported by {role.value}-relevant code"
    
    @staticmethod
    def _explain_context(
        file_path: str,
        role: Role,
        reason: str = ""
    ) -> str:
        """Explanation for CONTEXT priority files."""
        
        if reason:
            return f"Context file: {reason}"
        
        return f"Additional context - provides broader understanding for {role.value}"
    
    # ============================================
    # ROLE-SPECIFIC SUMMARIES
    # ============================================
    
    @staticmethod
    def generate_role_summary(
        file_content: str,
        file_path: str,
        role: Role,
        cached_summary: Optional[str] = None,
    ) -> str:
        """
        Generate role-specific summary of file.
        
        Args:
            file_content: File content (or summary)
            file_path: Path to file
            role: Target role
            cached_summary: Pre-computed summary (if available)
            
        Returns:
            Role-tailored summary
        """
        
        # Use cached if available
        if cached_summary:
            return ExplanationEngine._tailor_summary_to_role(cached_summary, role)
        
        # Generate summary based on file type
        file_ext = file_path.split('.')[-1].lower()
        
        if file_ext in ['tsx', 'jsx', 'vue']:
            return ExplanationEngine._summarize_component(file_content, role)
        elif file_ext in ['ts', 'js', 'py']:
            return ExplanationEngine._summarize_code(file_content, role)
        elif file_ext in ['yaml', 'yml', 'json']:
            return ExplanationEngine._summarize_config(file_content, role)
        elif file_ext in ['sql']:
            return ExplanationEngine._summarize_query(file_content, role)
        else:
            return "File content summary (role-specific view available)"
    
    @staticmethod
    def _summarize_component(content: str, role: Role) -> str:
        """Summarize a React/Vue component."""
        
        # Extract JSX/component structure
        functions = re.findall(r'function\s+(\w+)|const\s+(\w+)\s*=', content)
        hooks = re.findall(r'use\w+', content)
        props_match = re.search(r'interface\s+\w+Props|type\s+\w+Props', content)
        
        summary = "Component structure: "
        if functions:
            func_names = [f[0] or f[1] for f in functions]
            summary += f"exports {', '.join(func_names[:3])}. "
        
        if hooks:
            unique_hooks = list(set(hooks))
            summary += f"Uses hooks: {', '.join(unique_hooks[:3])}. "
        
        if role == Role.FRONTEND:
            summary += "This is a UI component. Key for rendering and user interaction."
        elif role == Role.BACKEND:
            summary += "Frontend component - references this from client-side code."
        
        return summary
    
    @staticmethod
    def _summarize_code(content: str, role: Role) -> str:
        """Summarize a code file."""
        
        # Extract exports/definitions
        exports = re.findall(r'export\s+(function|class|const|interface)\s+(\w+)', content)
        imports = re.findall(r'import.*?from\s+[\'"]([^\'"]+)[\'"]', content)
        
        summary = f"Defines: {len(exports)} exports. "
        if exports:
            export_names = [e[1] for e in exports[:3]]
            summary += f"Including {', '.join(export_names)}. "
        
        if imports:
            summary += f"Depends on {len(imports)} modules."
        
        if role == Role.BACKEND:
            if any(keyword in content.lower() for keyword in ['route', 'app.', 'handler']):
                summary += " Core API endpoint file."
        elif role == Role.FRONTEND:
            if any(keyword in content.lower() for keyword in ['component', 'render', 'jsx']):
                summary += " Component logic."
        
        return summary
    
    @staticmethod
    def _summarize_config(content: str, role: Role) -> str:
        """Summarize a config file (YAML/JSON)."""
        
        if role == Role.DEVOPS:
            if 'image' in content.lower():
                return "Deployment configuration - specifies container images and environment."
            elif 'resource' in content.lower():
                return "Infrastructure config - defines cloud resources."
            else:
                return "Configuration file - affects deployment environment."
        
        return "Configuration file - provides settings for application behavior."
    
    @staticmethod
    def _summarize_query(content: str, role: Role) -> str:
        """Summarize a SQL query."""
        
        if "CREATE TABLE" in content:
            tables = re.findall(r'CREATE TABLE\s+(\w+)', content)
            return f"Schema definition: creates table(s) {', '.join(tables)}"
        elif "SELECT" in content:
            return "Query file - retrieves data from database"
        elif "INSERT" in content or "UPDATE" in content:
            return "Mutation file - modifies database records"
        
        return "Database file - part of data access layer"
    
    @staticmethod
    def _tailor_summary_to_role(summary: str, role: Role) -> str:
        """Tailor an existing summary for a specific role."""
        
        # Add role-specific context to summary
        role_context = {
            Role.FRONTEND: " (Frontend impact: affects UI/UX)",
            Role.BACKEND: " (Backend impact: affects server-side logic)",
            Role.DEVOPS: " (DevOps impact: affects deployment)",
            Role.DATA: " (Data impact: affects data pipeline)",
            Role.SECURITY: " (Security impact: affects auth/encryption)",
        }
        
        if len(summary) > 100:
            return summary
        
        return summary + role_context.get(role, "")
    
    # ============================================
    # BATCH EXPLANATIONS
    # ============================================
    
    @staticmethod
    def explain_files_batch(
        files: List[Dict],  # {"file": path, "score": float, "role": Role}
        role: Role,
    ) -> List[Dict]:
        """
        Generate explanations for multiple files.
        
        Args:
            files: List of file metadata
            role: Target role
            
        Returns:
            Files with explanations added
        """
        results = []
        
        for file_info in files:
            explanation = ExplanationEngine.explain_file_relevance(
                file_path=file_info.get("file", ""),
                role=role,
                score=file_info.get("score", 0.5),
                primary_role=file_info.get("primary_role", "unknown"),
                reason=file_info.get("reason", ""),
            )
            
            file_info["explanation"] = explanation
            results.append(file_info)
        
        return results


def explain_file_relevance(
    file_path: str,
    role: Role,
    score: float,
) -> str:
    """
    Convenience function: explain file relevance for a role.
    
    Example:
        explanation = explain_file_relevance("api/controller.ts", Role.BACKEND, 0.95)
    """
    return ExplanationEngine.explain_file_relevance(
        file_path=file_path,
        role=role,
        score=score,
        primary_role="backend",
    )

File: src/explanationEngine.ts
/**
 * Explanation engine for generating role-specific file descriptions.
 * 
 * Generates contextual, data-driven explanations instead of generic templates.
 * Explains WHAT the file does, WHY it matters for the role, and WHERE it fits.
 */

import { Role } from "./config";

interface FileContext {
  path: string;
  role: Role;
  score: number;
  primaryRole: string;
  dependents?: string[]; // Files that depend on this
  dependencies?: string[]; // Files this depends on
}

export class ExplanationEngine {
  /**
   * Generate concise, specific explanation for why a file is shown for a role.
   * 
   * Strategy:
   * 1. Determine file's FUNCTION (what it does)
   * 2. Explain RELEVANCE to role (why it matters)
   * 3. Show POSITION in architecture (where it fits)
   */
  static explainFileRelevance(
    filePath: string,
    role: Role,
    score: number,
    primaryRole: string,
    reason: string = ""
  ): string {
    const fileType = filePath.split(".").pop()?.toLowerCase() || "";
    const fileName = filePath.split("/").pop() || filePath;
    const dirPath = filePath.includes("/") ? filePath.substring(0, filePath.lastIndexOf("/")) : "";

    // Use the reason if provided (from higher-level analysis)
    if (reason && reason.length > 5) {
      return reason;
    }

    // Determine file function first
    const fileFunction = ExplanationEngine.determineFileFunction(filePath, fileType);
    
    // Determine relevance level
    const priority = ExplanationEngine.determinePriority(score);
    
    // Build role-specific relevance
    const roleContext = ExplanationEngine.getRoleContext(role, fileFunction, primaryRole);

    // Combine into concise explanation
    return ExplanationEngine.buildExplanation(
      fileFunction,
      roleContext,
      priority,
      score
    );
  }

  /**
   * Determine what the file fundamentally does (not role-specific)
   */
  private static determineFileFunction(filePath: string, fileType: string): string {
    const lowerPath = filePath.toLowerCase();

    // API/Request Handling
    if (
      lowerPath.includes("controller") ||
      lowerPath.includes("handler") ||
      lowerPath.includes("route")
    ) {
      return "request_handler";
    }

    // UI Component
    if (
      lowerPath.includes("component") ||
      lowerPath.includes("page") ||
      lowerPath.includes("screen") ||
      fileType === "tsx" ||
      fileType === "jsx"
    ) {
      return "ui_component";
    }

    // Business Logic Service
    if (
      lowerPath.includes("service") &&
      !lowerPath.includes("auth")
    ) {
      return "business_service";
    }

    // Authentication/Authorization
    if (lowerPath.includes("auth")) {
      return "auth_service";
    }

    // Data Model/Schema
    if (
      lowerPath.includes("model") ||
      lowerPath.includes("schema") ||
      lowerPath.includes("entity") ||
      fileType === "sql"
    ) {
      return "data_model";
    }

    // Middleware
    if (lowerPath.includes("middleware")) {
      return "middleware";
    }

    // Configuration
    if (lowerPath.includes("config") || lowerPath.includes("settings")) {
      return "configuration";
    }

    // Utilities/Helpers
    if (
      lowerPath.includes("util") ||
      lowerPath.includes("helper") ||
      lowerPath.includes("lib")
    ) {
      return "utility";
    }

    // Testing
    if (
      lowerPath.includes("test") ||
      lowerPath.includes("spec") ||
      fileType === "spec"
    ) {
      return "test_suite";
    }

    // Infrastructure/Deployment
    if (
      fileType === "dockerfile" ||
      fileType === "yml" ||
      fileType === "yaml" ||
      lowerPath.includes("docker") ||
      lowerPath.includes("terraform") ||
      lowerPath.includes("helm")
    ) {
      return "infrastructure";
    }

    // Documentation
    if (fileType === "md" || lowerPath.includes("readme")) {
      return "documentation";
    }

    return "general_code";
  }

  /**
   * Determine priority level from score
   */
  private static determinePriority(
    score: number
  ): "primary" | "supporting" | "context" {
    if (score >= 0.7) return "primary";
    if (score >= 0.4) return "supporting";
    return "context";
  }

  /**
   * Get role-specific context for why file matters
   */
  private static getRoleContext(
    role: Role,
    fileFunction: string,
    primaryRole: string
  ): string {
    // Role doesn't match file's primary role
    if (role !== primaryRole) {
      return ExplanationEngine.getCrossRoleContext(role, fileFunction, primaryRole);
    }

    // Role matches - explain core relevance
    switch (fileFunction) {
      case "request_handler":
        return "Handles incoming requests and routing";
      case "ui_component":
        return "Renders UI and manages user interactions";
      case "business_service":
        return "Contains core business logic and operations";
      case "auth_service":
        return "Manages authentication and authorization";
      case "data_model":
        return "Defines data structures and schema";
      case "middleware":
        return "Intercepts and processes requests";
      case "configuration":
        return "Controls behavior and environment settings";
      case "utility":
        return "Provides shared helpers and utilities";
      case "test_suite":
        return "Validates functionality and correctness";
      case "infrastructure":
        return "Manages deployment and infrastructure";
      case "documentation":
        return "Explains features and usage";
      default:
        return "Contributes to core functionality";
    }
  }

  /**
   * Get context when role doesn't match file's primary role
   */
  private static getCrossRoleContext(
    role: Role,
    fileFunction: string,
    primaryRole: string
  ): string {
    // Frontend looking at backend
    if (role === Role.FRONTEND && primaryRole === Role.BACKEND) {
      if (fileFunction === "request_handler")
        return "Backend API that frontend depends on";
      if (fileFunction === "data_model") return "Data structure returned by API";
      if (fileFunction === "business_service") return "Business logic backend exposes";
      return "Backend code that supports frontend";
    }

    // Backend looking at frontend
    if (role === Role.BACKEND && primaryRole === Role.FRONTEND) {
      if (fileFunction === "ui_component")
        return "Frontend UI that calls your APIs";
      if (fileFunction === "request_handler")
        return "Frontend route you need to support";
      return "Frontend code that consumes your APIs";
    }

    // Anything looking at DevOps infrastructure
    if (role !== Role.DEVOPS && primaryRole === Role.DEVOPS) {
      return "Infrastructure requirements for this module";
    }

    // DevOps looking at application code
    if (role === Role.DEVOPS) {
      if (fileFunction === "infrastructure")
        return "Deployable infrastructure definition";
      if (fileFunction === "configuration") return "Environment configuration";
      return "Application code needing deployment";
    }

    // Security role looking at anything
    if (role === Role.SECURITY) {
      if (fileFunction === "auth_service")
        return "Authentication implementation to review";
      if (fileFunction === "request_handler")
        return "API endpoints that need security checks";
      if (fileFunction === "data_model")
        return "Data structures with potential sensitivities";
      return "Component with potential security implications";
    }

    return `Related to ${primaryRole} module`;
  }

  /**
   * Build final concise explanation
   */
  private static buildExplanation(
    fileFunction: string,
    roleContext: string,
    priority: string,
    score: number
  ): string {
    const scorePercent = (score * 100).toFixed(0);

    if (priority === "primary") {
      return `${roleContext} (${scorePercent}%) - essential for this role`;
    } else if (priority === "supporting") {
      return `${roleContext} (${scorePercent}%) - important dependency`;
    } else {
      return `${roleContext} (${scorePercent}%) - useful context`;
    }
  }

  /**
   * Get role-specific focus areas.
   */
  static getRoleFocus(role: Role | string): string {
    const focusMap: Record<string, string> = {
      frontend:
        "UI components, state management, styling, and user interactions",
      backend: "API endpoints, business logic, database queries, validation",
      devops: "deployment configs, CI/CD pipelines, infrastructure, monitoring",
      security: "authentication, authorization, encryption, access control",
      ai_ml: "model architecture, training logic, feature engineering",
      data: "data pipelines, ETL processes, database schemas",
      qa: "test coverage, test scenarios, validation logic",
      full_stack: "full data flow from UI to database, integration points",
    };

    return focusMap[role] || "general architecture";
  }

  /**
   * Generate a summary for a file based on role.
   */
  static summarizeFileForRole(
    fileContent: string,
    role: Role | string,
    filePath: string = ""
  ): string {
    const focus = this.getRoleFocus(role);

    return `Summarize this code for a ${role} engineer in 2-3 sentences.

Focus on ${focus}

File: ${filePath}
Code:
\`\`\`
${fileContent.substring(0, 2000)}
\`\`\`

Summary for ${role}:`;
  }

  /**
   * Explain a data flow path.
   */
  static explainDataFlow(
    path: string[],
    sourceFile: string,
    targetFile: string
  ): string {
    const pathStr = path.join(" → ");

    return `Explain this data flow path in one sentence:

${sourceFile} → ... → ${targetFile}

Path: ${pathStr}

Explanation:`;
  }
}
